### PR TITLE
Check if applied locale correctly lowers chars and fallback

### LINKED ISSUE #19883: Turkish locale set via LANG / LC_CTYPE / LC_ALL or by SetGlobalLocale breaks skin loading on any Linux
### Describe the bug

The following problem was reported to Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=989814

Having a generated `tr_TR.UTF-8` locale in the system breaks UI of **any** language resource, including English (see attached screenshots)

## Expected Behavior

Kodi UI is usable regardless of `LANG`, `LC_CTYPE` and `LC_ALL` environment variables set or system locales present in the system.

## Actual Behavior

The Turkish locale activated by either `SetGlobalLocale` call within Kodi or externally by means of `LANG`, `LC_CTYPE` and `LC_ALL` breaks UI

### To Reproduce

1. In the Debian system or container, ensure `tr_TR.UTF-8` locale is generated
2. Run in terminal:

``` shell
unset LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES \
      LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT \
      LC_IDENTIFICATION LC_ALL LANG LANGUAGE
locale
export LC_CTYPE="tr_TR.UTF-8"
kodi &
```

3. Experience unusable Kodi settings, movies, poweroff dialog etc.

### Debuglog

There are three debuglogs to explain the situation better:

1. [kodi-uz_UZ-resource-without-uz_UZ.UTF-8-locale.txt](https://github.com/xbmc/xbmc/files/6651147/kodi-uz_UZ-resource-without-uz_UZ.UTF-8-locale.txt) - Kodi correctly operates with Uzbek language resource (uz_UZ) and *no* Uzbek locale `uz_UZ.UTF-8` is generated in Debian system
2. [kodi-uk_UA-resource-with-uk_UA.UTF-8-locale.txt](https://github.com/xbmc/xbmc/files/6651152/kodi-uk_UA-resource-with-uk_UA.UTF-8-locale.txt) - Kodi correctly operates with Ukrainian language resource (uk_UA) and Ukrainian locale `uk_UA.UTF-8` *is present* in Debian system
3. [kodi-installing-tr_TR-resource-with-tr_TR.UTF-8-locale.txt](https://github.com/xbmc/xbmc/files/6651159/kodi-installing-tr_TR-resource-with-tr_TR.UTF-8-locale.txt) - Kodi is broken if `LC_CTYPE` is set to `tr_TR.UTF-8` and Kodi starts with Ukrainian language and then user tries to download Turkish language and Kodi becomes completely unusable.

### Screenshots 
Here are some links or screenshots to help explain the problem:

uk_UA-resource-on-LC_CTYPE-tr_TR UTF-8-01
uk_UA-resource-on-LC_CTYPE-tr_TR UTF-8-02

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

The first screenshot is companion to log number 3. The rest two are showing broken Kodi UI on any language if `LC_CTYPE` is set to Turkish locale.

### Your Environment
Used Operating system:
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Debian 11 bullseye
 - Kodi version: 2:19.1+dfsg2-1

### PR DESCRIPTION
**Pull request by basilgello**
.. to default region if it does not.

Fixes #19883.

### CODE REVIEW

**Comment by basilgello** [comment]
Got an ACK from @hbayindir in the Debian bug report thread:

```
Thanks a lot for your intense effort. The new version you uploaded is working 
as intended without any changes to locale.

I can confirm that the problem is now solved.
```

Let's start the review?

**Comment by basilgello** [comment]
@fuzzard @lrusak @Rechi @phunkyfish @DaveTBlake Sorry I dont see who is actually code owner of `LangInfo.cpp` - it was written so long ago…  :)

File: xbmc/LangInfo.cpp
Comment (lrusak) [inline xbmc/LangInfo.cpp:986]:
```suggestion
  if (test19883 != 0)
```

File: xbmc/LangInfo.cpp
Comment (lrusak) [inline xbmc/LangInfo.cpp:988]: Should probably log the locale that caused the error and what the default locale is.

File: xbmc/LangInfo.cpp
Comment (lrusak) [inline xbmc/LangInfo.cpp:985]:
```suggestion
  int test19883 = std::tolower('i') - std::::tolower('I');
```

**Comment by basilgello** [comment]
@lrusak Fixed, thanks! Not sure if `m_defaultRegion.m_strName` is meaningful though, but at least the failing region name should be displayed correctly and the latter messages from `SetGlobalLocale` show the locale is set to `C`

**Comment by basilgello** [comment]
> Not sure if m_defaultRegion.m_strName is meaningful though

Yes, default region is marked as 'N/A' but the locale info is displayed as INFO:

```
2021-06-30 05:11:15.500 T:8514     INFO <general>: CAddonMgr::FindAddon: resource.language.tr_tr v9.0.22 installed
2021-06-30 05:11:15.551 T:8514     INFO <general>: CLangInfo: loading resource.language.tr_tr language information...
2021-06-30 05:11:15.552 T:8514     INFO <general>: global locale set to tr_TR.UTF-8
2021-06-30 05:11:15.552 T:8514    ERROR <general>: region 'Türkiye' is affected by #19883 - falling back to default region 'N/A'
2021-06-30 05:11:15.552 T:8514     INFO <general>: global locale set to C
2021-06-30 05:11:15.552 T:8514     INFO <general>: CLangInfo: loading resource.language.tr_tr language strings...
2021-06-30 05:11:15.597 T:8514     INFO <general>: global locale set to tr_TR.UTF-8
2021-06-30 05:11:15.597 T:8514    ERROR <general>: region 'Türkiye' is affected by #19883 - falling back to default region 'N/A'
2021-06-30 05:11:15.597 T:8514     INFO <general>: global locale set to C
```

I think there's no need to move ifdeffery from https://github.com/xbmc/xbmc/blob/master/xbmc/LangInfo.cpp#L222 into a separate function  `std::string CLangInfo::GetLocaleName(CLangInfo::CRegion& region);` just to print exact locale names?

**Comment by lrusak** [comment]
I think it's fine as is. The only other thing I can say is maybe change to `LOGWARNING` rather than `LOGERROR` but I'm fine either way.

**Comment by phunkyfish** [comment]
jenkins build this please

File: tools/Linux/kodi.sh.in
Comment (howie-f) [inline tools/Linux/kodi.sh.in:176]: do we need to unset those in further files - example https://github.com/xbmc/xbmc/blob/master/tools/Linux/kodi-standalone.sh.in ? and what are consequences of unsetting as LC_ALL is saved and restored here https://github.com/xbmc/xbmc/blob/8d098aee94781c9fad5c383f69242a69984bb69a/xbmc/windowing/X11/WinEventsX11.cpp#L175-L204

just an informational question, i'm not really deep into these things

File: xbmc/LangInfo.cpp
Comment (howie-f) [inline xbmc/LangInfo.cpp:986]: maybe change to
```c
const bool fallbackToDefaultRegionNeeded = (std::tolower('i') - std::tolower('I')) != 0;
if (fallbackToDefaultRegionNeeded)
```
plus a little comment/description why this check is needed would be awesome

**Review by howie-f — COMMENTED** [review]
i did not see this bug on windows 10 with turkish language (system and kodi) installed.
unfortunately can't test android and osx... android would be interesting though

File: xbmc/LangInfo.cpp
Comment (garbear) [reply xbmc/LangInfo.cpp:986]: You can add `std::tolower('I')` to both sides and compare `(std::tolower('i') != std::tolower('I'));`

**Comment by basilgello** [comment]
> android would be interesting though

I just tested 19.1 from F-Droid in my LineageOS tablet and it seems Android is not vulnerable due to https://github.com/xbmc/xbmc/blob/19.1-Matrix/xbmc/LangInfo.cpp#L297

In fact, `C.UTF-8` becomes the default region applied by this fix if `Turkish I` is detected.

> do we need to unset those in further files - example https://github.com/xbmc/xbmc/blob/master/tools/Linux/kodi-standalone.sh.in ? and what are consequences of unsetting as LC_ALL

No, we need to unset `LC_*` variables only once, just before `kodi.bin` is invoked. Unsetting and setting back `LC_ALL` in X11 code alters the already-unset value so it should not interfere. The proper way avoiding `kodi.sh.in` edit is to understand what Kodi dependency breaks rendering before `GetLocaleName` is first invoked, but I did not manage to dig that out.

**Comment by basilgello** [comment]
@phunkyfish can you please run unpatched 19.1 on OSX with Turkish system and Kodi language and check whether the bug manifests?

**Comment by phunkyfish** [comment]
> @phunkyfish can you please run unpatched 19.1 on OSX with Turkish system and Kodi language and check whether the bug manifests?

How exactly do I do that? Being an English speaker I never deal with other langauges!!!

**Comment by basilgello** [comment]
Temporary set system language to Turkish (you can leave region intact!) and start Kodi, then install Turkish language via Settings - Interface - Region and language then restart Kodi and check if movie windows, poweroff prompt etc are readable.

**Comment by phunkyfish** [comment]
How will I know if they are readable? If there is a screenshot of what they should look like?

**Comment by basilgello** [comment]
Screenshots of affected Kodi are in #19883

**Comment by fuzzard** [comment]
Screen Shot 2021-09-13 at 9 14 51 am

Appears to be ok (if ive done what you have said properly. Was just a stock 19.1 release build, MacOS 11.5.2

**Comment by basilgello** [comment]
Yes, thanks! So it is Linux-only issue!

**Comment by howie-f** [comment]
should be ready after it got a bit love and polishing 😉

**Comment by basilgello** [comment]
Yeah, I paused this after I tried to move `kodi.sh.in` part into `xbmc/platform/posix/main.cpp` and the fix failed to address an issue :) Gonna try it out once more & fix the rest + backport!

**Comment by basilgello** [comment]
@howie-f I fixed the points you brought on the last review! Unfortunately I could not get rid of un-setting environment variables in `kodi.sh.in` because **something** does not honor even the thermonuclear locale setting:

```cpp
   std::setlocale(LC_ALL, "C");
   std::locale::global(std::locale("C"));
```

I think we should go with battle-tested code here, and the fact that `kodi.sh.in` is used only by Linux platform suggest it is a correct place to alter `LC_*` stuff.

**Comment by ksooo** [comment]
Sorry, I don't wont to be PITA, but shouldn't we try to find and fix the root cause of this problem instead of polluting Kodi code base with workarounds?

**Comment by basilgello** [comment]
> Sorry, I don't want to be PITA, but shouldn't we try to find and fix the root cause of this problem instead of polluting Kodi code base with workarounds?

One of root cases of this problem is that https://github.com/xbmc/xbmc/blob/19.1-Matrix/xbmc/filesystem/SpecialProtocol.cpp#L207 uses lower/upper functions to find files in a case-insensitive manner. There are others + some that even I failed to figure out by today :(

I am OK to solve that issue completely, but I am really afraid (see my comments in 19883) that the 'true' fix will be greater than CDateTime one :(

**Comment by ksooo** [comment]
> but I am really afraid (see my comments in 19883) that the 'true' fix will be greater than CDateTime one :(

 Let's do the job right and bite that bullet. 
 
  Kodi "survived" without this fix for many many years. I don't see the need for this kinda special case to hurry and go with a workaround.

**Comment by basilgello** [comment]
> Kodi "survived" without this fix for many many years. I don't see the need for this kinda special case to hurry and go with a workaround.

Matrix users from Turkey already reported this on r/kodi and in Debian. I am OK to hold this for Nexus, but Matrix should be worked around. I do ship the fix already in Debian, but Manjaro / Arch / Gentoo users still lack it. Should we leave it as-is in such case?

**Review by ksooo — CHANGES_REQUESTED** [review]
Sorry, I don't see this in master. This is only a workaround and we know that the root cause is in Kodi code base. So, let's fix it properly.

**Comment by ksooo** [comment]
> Matrix users from Turkey already reported this on r/kodi and in Debian. I am OK to hold this for Nexus, but Matrix should be worked around.

I do agree.

**Pull request closed without merging**
